File: Java/simple/13.罗马数字转整数.java
package Java.simple;

import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;
// 定义一个HashMap，然后判断当前字符是否小于下一个字符，如果是，就减去当前字符的值，否则加上当前字符的值
class Solution {
    Map<Character, Integer> symbolValues = new HashMap<Character, Integer>() {{
        put('I', 1);
        put('V', 5);
        put('X', 10);
        put('L', 50);
        put('C', 100);
        put('D', 500);
        put('M', 1000);
    }};

    public int romanToInt(String s) {
        int ans = 0;
        int n = s.length();
        for (int i = 0; i < n; ++i) {
            int value = symbolValues.get(s.charAt(i));
            if (i < n - 1 && value < symbolValues.get(s.charAt(i + 1))) {
                ans -= value;
            } else {
                ans += value;
            }
        }
        return ans;
    }


    public static void main(String[] args) {
        Solution solution = new Solution();
        int ans = solution.romanToInt("III");
        System.out.println(ans);
        ans = solution.romanToInt("IV");
        System.out.println(ans);
        ans = solution.romanToInt("IX");
        System.out.println(ans);
        ans = solution.romanToInt("LVIII");
        System.out.println(ans);
        ans = solution.romanToInt("MCMXCIV");
        System.out.println(ans);
        Object[] array = solution.symbolValues.entrySet().toArray();
        System.err.println(Arrays.toString(array));    // [C=100, D=500, V=5, X=10, I=1, L=50, M=1000]
    }
}

File: Kotlin/simple/13.罗马数字转整数.kt
class Solution {
    private fun getValue(c:Char):Int = when(c) {
        'I' -> 1
        'V' -> 5
        'X' -> 10
        'L' -> 50
        'C' -> 100
        'D' -> 500
        'M' -> 1000
        else -> throw UnsupportedOperationException("不支持的字符: $c")
    }


    fun romanToInt(s: String): Int {
        var result = 0
        var prevValue = 0
        
        s.forEach { char ->
            val currentValue = getValue(char)
            if (currentValue > prevValue) {
                // 当前值大于前一个值，说明是特殊情况（如IV、IX等）
                // 需要减去之前加过的前一个值，然后加上（当前值-前一个值）
                result += currentValue - 2 * prevValue
            } else {
                // 正常情况，直接加上当前值
                result += currentValue
            }
            prevValue = currentValue
        }
        
        return result
    }
}
// fun main{
// print(Solution().romanToInt("III"))
// print(Solution().romanToInt("IV"))
// print(Solution().romanToInt("IX"))
// print(Solution().romanToInt("LVIII"))
// print(Solution().romanToInt("MCMXCIV"))
// }

val result = Solution().romanToInt("III")
print(result)